### ISSUE TITLE
[7.0] Timesheet : Timezone error for total attendances

### ISSUE THREAD

**Issue by vrenaville** [odoo/odoo#5379]
Timezone is alway UTC , when you need to compute delta 
between current_date and last sign in of the user.

https://github.com/odoo/odoo/pull/5378

**Comment by gurneyalex** [comment]
git blame suggests this issue was probably introduced in dbb2a669f44eff8889d96e0ed5113fd394ae1cae  (but I did not bisect it) cc @beledouxdenis

**Comment by beledouxdenis** [comment]
The employee of the timesheet needs to have set its timezone in order this field to work correctly. Is it the case ?

**Comment by gurneyalex** [comment]
I had the problem here at camptocamp, and my user has the timezone set properly.

**Comment by beledouxdenis** [comment]
Is it possible to describe exactly the steps needed to reproduce this issue ?

**Comment by gurneyalex** [comment]
1. connect to openerp
2. in settings enable "track attendance"
3. Check that your user is in Europe / Brussels timezone. 
4. Create a new timesheet, add a sign in attendance line through the form view (by clicking on sign in in the attendance tab of the form)

Expected: the total attendance in the top part of the form is 0:00

Actual: the total attendance in the top part of the form in -1:00

**Comment by gurneyalex** [comment]
I reproduce also by signing in using the widget in the top right corner of the screen. 

I reproduce also by signing it by manually inputing an attendance line in the appropriate o2m field of the timesheet form.

**Comment by beledouxdenis** [comment]
When I sign in, I indeed have -1, which is not correct.

Nevertheless, as soon as I enter a sign out, the amount is correct.

So, the issue only appears when sign in is not coupled with its sign out, am I right ?

**Comment by gurneyalex** [comment]
yes, and this is very annoying because 

1. people like to fill in there timesheet activities before signing out, and at that time the total attendance displayed is incorrect

2. it gives a poor impression for people living a place not using UTC as their timezone

3. it makes me wonder what happen if I live in UTC+8 and sign in at 7AM local time on monday: does the attendance even land in the correct timesheet?

**Comment by beledouxdenis** [comment]
I didn't say I wasn't going to fix it, I was just describing in which states the issue occurs.

I am on it right now.

3.it makes me wonder what happen if I live in UTC+8 and sign in at 7AM local time on monday: does the attendance even land in the correct timesheet?

This is exactly the point of my fix you mentioned. Before, the employee timezone was just ignored.

**Comment by gurneyalex** [comment]
Sorry I had misinterpreted your question. 

Thanks for looking into this :smile_cat:

**Comment by beledouxdenis** [comment]
Except the fact the sub select is called 'foo', the pull request seems the right fix to me :).

**Comment by vrenaville** [comment]
@beledouxdenis sub select is called 'foo' before my commit, so I preserve it ;) , but I can rename it (of course) if request

**Comment by beledouxdenis** [comment]
Well, my bad, didn't saw it :x.

I am still reviewing it. I come back within an hour.

**Comment by gurneyalex** [comment]
Just did a bit of archive digging... 

foo was introduced in https://github.com/odoo/odoo/commit/bacfe373 and bar came along a few days later in https://github.com/odoo/odoo/commit/89b31248

**Comment by beledouxdenis** [comment]
The patch is fine, but I am just trying to find a way to avoid the two big joins. I believe only one is possible, I keep looking.

**Comment by beledouxdenis** [comment]
There : )
https://github.com/odoo-dev/odoo/compare/odoo:7.0...7.0-fixworkedhourssignin-dle?expand=1

Isn't this cleaner ?  : D.

**Comment by gurneyalex** [comment]
are the foreign key columns indexed? If so, the joins should not be too costly...

**Comment by gurneyalex** [comment]
Cleaner? I don't know... 

Simpler? probably (although maybe it's time to rename `foo` and `bar` after all these years :wink:)

Faster? I'd like to see an EXPLAIN ANALYZE on a loaded and vacuumed database. 

Correct? can I talk you into adding an automated test for this which will save the next guy wanting to meddle with this query to introduce regressions? :smile_cat:

**Comment by beledouxdenis** [comment]
I agree with you that a test is needed.

But, almost all my fixes would require a test along with them, but If I did so, It would take me like two times more time, or even more. 

You will probably argue that at long term it will save time as regressions will be avoided, and I agree with you. Nevertheless, in my current environment, customers just ask to repair issues the fastest possible, and I try to do so. I should probably have a guy next to me writing tests for every patch I made. I can still try to ask the head for it : D.

**Comment by gurneyalex** [comment]
@vrenaville  version on Camptocamp database:

    explain analyze select * from hr_timesheet_sheet_sheet_day where sheet_id=16158;
                                                                                                 QUERY PLAN                                                                                             
    ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
     Subquery Scan on bar  (cost=12583.85..12585.67 rows=26 width=28) (actual time=137.404..137.476 rows=4 loops=1)
       ->  HashAggregate  (cost=12583.85..12585.15 rows=26 width=28) (actual time=137.368..137.426 rows=4 loops=1)
             ->  HashAggregate  (cost=12582.88..12583.14 rows=26 width=21) (actual time=137.349..137.351 rows=4 loops=1)
                   ->  Append  (cost=6289.16..12582.55 rows=26 width=21) (actual time=137.329..137.337 rows=4 loops=1)
                         ->  Subquery Scan on "*SELECT* 1"  (cost=6289.16..6289.54 rows=19 width=20) (actual time=51.782..51.782 rows=0 loops=1)
                               ->  HashAggregate  (cost=6289.16..6289.35 rows=19 width=20) (actual time=51.781..51.781 rows=0 loops=1)
                                     ->  Nested Loop  (cost=0.00..6288.97 rows=19 width=20) (actual time=51.777..51.777 rows=0 loops=1)
                                           ->  Index Scan using hr_timesheet_sheet_sheet_pkey on hr_timesheet_sheet_sheet s  (cost=0.00..8.27 rows=1 width=4) (actual time=0.046..0.053 rows=1 loops=1)
                                                 Index Cond: (id = 16158)
                                           ->  Nested Loop  (cost=0.00..6280.51 rows=19 width=20) (actual time=51.716..51.716 rows=0 loops=1)
                                                 ->  Seq Scan on hr_analytic_timesheet hrt  (cost=0.00..6120.98 rows=19 width=12) (actual time=51.714..51.714 rows=0 loops=1)
                                                       Filter: (sheet_id = 16158)
                                                 ->  Index Scan using account_analytic_line_pkey on account_analytic_line l  (cost=0.00..8.38 rows=1 width=16) (never executed)
                                                       Index Cond: (id = hrt.line_id)
                         ->  Subquery Scan on "*SELECT* 2"  (cost=6292.84..6293.01 rows=7 width=24) (actual time=85.545..85.550 rows=4 loops=1)
                               ->  HashAggregate  (cost=6292.84..6292.94 rows=7 width=24) (actual time=85.527..85.530 rows=4 loops=1)
                                     ->  Nested Loop  (cost=0.00..6292.56 rows=10 width=24) (actual time=75.201..85.468 rows=9 loops=1)
                                           ->  Index Scan using hr_timesheet_sheet_sheet_pkey on hr_timesheet_sheet_sheet s  (cost=0.00..8.27 rows=1 width=4) (actual time=0.020..0.027 rows=1 loops=1)
                                                 Index Cond: (id = 16158)
                                           ->  Seq Scan on hr_attendance a  (cost=0.00..6284.17 rows=10 width=24) (actual time=75.157..85.405 rows=9 loops=1)
                                                 Filter: (((action)::text = ANY ('{sign_in,sign_out}'::text[])) AND (sheet_id = 16158))
     Total runtime: 138.568 ms
    (22 rows)


@beledouxdenis denis version on same database:

    # explain analyze select * from hr_timesheet_sheet_sheet_day2 where sheet_id=16158;
                                                                                                                QUERY PLAN                                                                                                            
    ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
     Subquery Scan on bar  (cost=12148.69..12150.94 rows=30 width=28) (actual time=195.838..195.882 rows=4 loops=1)
       ->  HashAggregate  (cost=12148.69..12150.34 rows=30 width=60) (actual time=195.812..195.844 rows=4 loops=1)
             ->  HashAggregate  (cost=12147.49..12147.79 rows=30 width=27) (actual time=195.791..195.793 rows=4 loops=1)
                   ->  Append  (cost=5749.66..12147.04 rows=30 width=27) (actual time=195.772..195.781 rows=4 loops=1)
                         ->  Subquery Scan on "*SELECT* 1"  (cost=5749.66..5750.02 rows=18 width=20) (actual time=72.772..72.772 rows=0 loops=1)
                               ->  HashAggregate  (cost=5749.66..5749.84 rows=18 width=20) (actual time=72.771..72.771 rows=0 loops=1)
                                     ->  Nested Loop  (cost=0.00..5749.44 rows=18 width=20) (actual time=72.768..72.768 rows=0 loops=1)
                                           ->  Index Scan using hr_timesheet_sheet_sheet_pkey on hr_timesheet_sheet_sheet s  (cost=0.00..8.27 rows=1 width=4) (actual time=0.043..0.052 rows=1 loops=1)
                                                 Index Cond: (id = 16158)
                                           ->  Nested Loop  (cost=0.00..5740.99 rows=18 width=20) (actual time=72.709..72.709 rows=0 loops=1)
                                                 ->  Seq Scan on hr_analytic_timesheet hrt  (cost=0.00..5589.85 rows=18 width=12) (actual time=72.708..72.708 rows=0 loops=1)
                                                       Filter: (sheet_id = 16158)
                                                 ->  Index Scan using account_analytic_line_pkey on account_analytic_line l  (cost=0.00..8.38 rows=1 width=16) (never executed)
                                                       Index Cond: (id = hrt.line_id)
                         ->  Subquery Scan on "*SELECT* 2"  (cost=6396.66..6397.02 rows=12 width=37) (actual time=122.997..123.002 rows=4 loops=1)
                               ->  HashAggregate  (cost=6396.66..6396.90 rows=12 width=37) (actual time=122.978..122.981 rows=4 loops=1)
                                     ->  Nested Loop Left Join  (cost=0.00..6396.18 rows=12 width=37) (actual time=122.492..122.906 rows=9 loops=1)
                                           ->  Nested Loop Left Join  (cost=0.00..6321.11 rows=12 width=28) (actual time=122.408..122.772 rows=9 loops=1)
                                                 ->  Nested Loop  (cost=0.00..6312.53 rows=12 width=28) (actual time=122.387..122.719 rows=9 loops=1)
                                                       ->  Nested Loop  (cost=0.00..6307.11 rows=12 width=28) (actual time=122.364..122.664 rows=9 loops=1)
                                                             Join Filter: (a.employee_id = e.id)
                                                             ->  Seq Scan on hr_employee e  (cost=0.00..4.93 rows=93 width=8) (actual time=0.022..0.064 rows=93 loops=1)
                                                             ->  Materialize  (cost=0.00..6285.47 rows=12 width=28) (actual time=1.206..1.315 rows=9 loops=93)
                                                                   ->  Nested Loop  (cost=0.00..6285.41 rows=12 width=28) (actual time=112.096..122.081 rows=9 loops=1)
                                                                         ->  Index Scan using hr_timesheet_sheet_sheet_pkey on hr_timesheet_sheet_sheet s  (cost=0.00..8.27 rows=1 width=4) (actual time=0.022..0.028 rows=1 loops=1)
                                                                               Index Cond: (id = 16158)
                                                                         ->  Seq Scan on hr_attendance a  (cost=0.00..6277.02 rows=12 width=28) (actual time=112.059..122.032 rows=9 loops=1)
                                                                               Filter: (((action)::text = ANY ('{sign_in,sign_out}'::text[])) AND (sheet_id = 16158))
                                                       ->  Index Scan using resource_resource_pkey on resource_resource r  (cost=0.00..0.44 rows=1 width=8) (actual time=0.003..0.004 rows=1 loops=9)
                                                             Index Cond: (id = e.resource_id)
                                                 ->  Index Scan using res_users_pkey on res_users u  (cost=0.00..0.70 rows=1 width=8) (actual time=0.003..0.003 rows=1 loops=9)
                                                       Index Cond: (r.user_id = id)
                                           ->  Index Scan using res_partner_pkey on res_partner p  (cost=0.00..6.24 rows=1 width=17) (actual time=0.004..0.005 rows=1 loops=9)
                                                 Index Cond: (u.partner_id = id)
     Total runtime: 196.202 ms
    (35 rows)

**Comment by beledouxdenis** [comment]
Well, if this is your production database, it depends on the current load and drive access at the exact time you launch you request.

But, honestly, I just don't care, I said from the beginning that @vrenaville is correct, and was just looking if there wasn't a cleaner / faster way. I will merge #5378.

**Comment by odony** [comment]
The thing is, analyzing EXPLAIN PLANS is not trivial. For starters, you can't compare 2 queries through a single execution time - the variance is too high: it depends on a lot of things like the current activity on your machine, I/O caused by other processes, state of your PG and OS cache, etc. With query times under 200ms the variance is so high you can't compare anything without collecting real statistics over thousands of runs, and with larger volumes.
And that would be an exercise in futility.

Based on the theoretical cost of your samples (`cost=12583.85..12585.67` vs `cost=12148.69..12150.94`), Denis' version appears to be slightly faster, but only by a few percents. That's low enough to be negligible in practice.

But even *that* theoretical comparison is not reliable. It is only valid for your current database, with your current data, and your current statistics. And it could be completely wrong **even for you** if your statistics are outdated (running `ANALYZE` could significantly alter your query plans, YMMV).

Bottom line: the speed difference is negligible and simpler code is always preferable in that case.

**Comment by odony** [comment]
@vrenaville Please do not open Issues when you are posting Pull Requests. A Pull Request **is** an issue, just post the issue details in the description of the PR. That means half the work for you and half the work for us. It is now [the first rule in the contributing guide](https://github.com/odoo/odoo/blob/8.0/CONTRIBUTING.md). Thanks!

**Comment by gurneyalex** [comment]
@odony you are correct about the execution time. I did a vacuum first though so the query plans are not too widely off.

**Comment by vrenaville** [comment]
@beledouxdenis 
Hello do you need more information about my PR ? 

Thanks,

Vincent

**Comment by beledouxdenis** [comment]
Hello, 

After analysis, we will merge this version:
https://github.com/odoo-dev/odoo/compare/odoo:7.0...7.0-fixworkedhourssignin-dle?expand=1

I can nevertheless set you as author, if you want, as you did most part of the job.

Regards,

**Comment by vrenaville** [comment]
@beledouxdenis Thanks for your respond.
could you please make a PR for your fix ?
Thanks for your help 

Vincent

**Comment by beledouxdenis** [comment]
Hello @vrenaville,

I spotted another issue when the employee set midnight exactly as sign in and no sign out: No worked hours were displayed.

I updated my branch with a new commit. See PR #5503.

I will probably merge this today, if everything is fine for you.

Regards,

**Comment by vrenaville** [comment]
Hello,
@beledouxdenis 
I have made a test, and seems good to me except the foo / bar query name :tongue: 

Vincent

**Comment by beledouxdenis** [comment]
Ahah good one :p. I will proceed to the merge in stable release 7.0.

**Comment by sylvain-garancher** [comment]
Hello,

We have exactly the same error in 8.0.
Should I create another issue ?

**Comment by odony** [comment]
@sylvain-garancher Fixes done in 7.0 will be automatically ported to 8.0, so you don't need to do anything :-)

**Comment by vrenaville** [comment]
@beledouxdenis Could you please check the Issue #5809 , we are facing a problem if you made multiple sign/in sign/out.
I have test my previous PR on this issue and it's work ;) could you please avec a look ?
